[JSC] Add fast paths for global atom RegExp matching one char and multiple chars

hyjorc1 opened the pull request:
#### 408cc4c30591c741bb1ae14bebd1f060eb62619b
<pre>
[JSC] Add fast paths for global atom RegExp matching one char and multiple chars
<a href="https://bugs.webkit.org/show_bug.cgi?id=281265">https://bugs.webkit.org/show_bug.cgi?id=281265</a>
<a href="https://rdar.apple.com/137717054">rdar://137717054</a>

Reviewed by Yusuke Suzuki.

Inspired by [1], this patch add two fast paths to let global
atom RegExp to match strings quickly with simple string search.

[1] <a href="https://chromium-review.googlesource.com/c/v8/v8/+/5872644">https://chromium-review.googlesource.com/c/v8/v8/+/5872644</a>

                                         before                    after

global-atom-substring-match          2.2508+-0.0145     ^      1.6873+-0.0077        ^ definitely 1.3340x faster
global-atom-match-multi-chars        2.4767+-0.0195     ^      1.5488+-0.0071        ^ definitely 1.5991x faster
global-atom-match-one-char           8.0799+-0.0309     ^      1.2977+-0.0078        ^ definitely 6.2264x faster

* JSTests/microbenchmarks/global-atom-match-multi-chars.js: Copied from JSTests/microbenchmarks/substring-match-global-atom.js.
(test):
* JSTests/microbenchmarks/global-atom-match-one-char.js: Copied from JSTests/microbenchmarks/substring-match-global-atom.js.
(test):
* JSTests/microbenchmarks/global-atom-substring-match.js: Renamed from JSTests/microbenchmarks/substring-match-global-atom.js.
(test):
* JSTests/stress/string-16bit-repeat-overflow.js:
* JSTests/stress/substring-global-atom-cache.js:
(test):
* Source/JavaScriptCore/runtime/RegExpObject.cpp:
(JSC::RegExpObject::matchGlobal):
* Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
(JSC::GlobalAtomMeta::isSubstring):
(JSC::oneCharMatches):
(JSC::genericMatches):
(JSC::collectGlobalAtomMatches):
* Source/JavaScriptCore/runtime/RegExpSubstringGlobalAtomCache.cpp:
(JSC::RegExpSubstringGlobalAtomCache::tryGet):
(JSC::RegExpSubstringGlobalAtomCache::tryRecord):
(JSC::RegExpSubstringGlobalAtomCache::hasValidPattern): Deleted.
(JSC::RegExpSubstringGlobalAtomCache::collectMatches): Deleted.
* Source/JavaScriptCore/runtime/RegExpSubstringGlobalAtomCache.h:

Canonical link: <a href="https://commits.webkit.org/285090@main">https://commits.webkit.org/285090@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ba7d126373f280c6bc17d6fe47d17a0e3b66eb7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71533 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50946 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24307 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75645 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22738 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58746 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22558 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/14967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74599 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/46224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/61615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/36944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/42887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/19079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21079 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64658 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/64790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/19445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77363 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70782 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15767 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15810 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/61654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/12351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/5989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92568 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10961 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46746 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/20410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/47817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49101 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47559 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
```diff
@@ -225,4 +226,101 @@ JSValue collectMatches(VM& vm, JSGlobalObject* globalObject, JSString* string, c
+inline void oneCharMatches(std::span<const Char> input, Char pattern, GlobalAtomMeta& meta)
+{
+    for (size_t i = meta.startIndex; i < input.size(); ++i) {
+        if (input[i] != pattern)
```
How about using WTF::find? It is SIMD accelerated. Let's try and see the difference (if the input is small, it is possible that it becomes a bit slower)

hyjorc1 replied:
`before`: current version.
`after`: `WTF::find` version.
```
                                      before                    after                                       

global-atom-match-one-char        2.7126+-0.0087     !      2.9283+-0.0082        ! definitely 1.0795x slower
```
Yeah, it seems `WTF::find` is slower with this
```
    size_t found;
    while ((found = WTF::find(input, pattern, meta.startIndex)) != notFound) {
        meta.numberOfMatches++;
        meta.startIndex = found + 1;
        if (meta.numberOfMatches > MAX_STORAGE_VECTOR_LENGTH)
            return;
    }
```

Constellation commented on Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
```diff
@@ -225,4 +226,101 @@ JSValue collectMatches(VM& vm, JSGlobalObject* globalObject, JSString* string, c
+    }
+}
+
+inline void genericMatches(StringView input, StringView pattern, GlobalAtomMeta& meta)
```
How about using span too here and using WTF::find?

hyjorc1 replied:
It seems `WTF::find` doesn't support searching characters. Btw, `StringView::find` is based on `findInner` which ready uses SIMD based `equal` checks.

Constellation commented on Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
```diff
@@ -225,4 +226,101 @@ JSValue collectMatches(VM& vm, JSGlobalObject* globalObject, JSString* string, c
+    }
+
+    // Construct the array
+    JSArray* array = constructEmptyArray(globalObject, nullptr, meta.numberOfMatches);
```
Let's consider doing super fast allocation here for major case.
1. First, get the pattern string
2. Next, try allocate Butterfly with the size (if it is not possible (due to slow array structure etc., use the current slow code as a fallback)
3. Fill the butterfly with the (1)'s pointer. You can use memset_pattern8 on Darwin further
4. Create array with the butterfly.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -319,6 +319,56 @@ inline JSArray* constructArrayNegativeIndexed(JSGlobalObject* globalObject, Arra
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    Structure* structure = globalObject->originalArrayStructureForIndexingType(ArrayWithContiguous);
+    RETURN_IF_EXCEPTION(scope, { });
```
It may say that this structure is not Contiguous (if having a bad time is happening). In that case, we should use the slow path.

hyjorc1 replied:
Thanks, updated. Just want to confirm that the update is what you were expecting.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -319,6 +319,56 @@ inline JSArray* constructArrayNegativeIndexed(JSGlobalObject* globalObject, Arra
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    Structure* structure = globalObject->originalArrayStructureForIndexingType(ArrayWithContiguous);
+    RETURN_IF_EXCEPTION(scope, { });
```
You do not need to have `RETURN_IF_EXCEPTION` for `originalArrayStructureForIndexingType`.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -319,6 +319,56 @@ inline JSArray* constructArrayNegativeIndexed(JSGlobalObject* globalObject, Arra
+        return nullptr;
+
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
```
So, you do not need this scope too.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -319,6 +319,53 @@ inline JSArray* constructArrayNegativeIndexed(JSGlobalObject* globalObject, Arra
+        vm,
+        Butterfly::totalSize(0, 0, true, vectorLength * sizeof(EncodedJSValue)),
+        nullptr, AllocationFailureMode::ReturnNull);
+    if (!temp)
```
UNLIKELY

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -319,6 +319,53 @@ inline JSArray* constructArrayNegativeIndexed(JSGlobalObject* globalObject, Arra
+
+    VM& vm = globalObject->vm();
+    Structure* structure = globalObject->originalArrayStructureForIndexingType(ArrayWithContiguous);
+    if (!hasContiguous(structure->indexingType()))
```
UNLIKELY

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
```diff
@@ -319,6 +319,53 @@ inline JSArray* constructArrayNegativeIndexed(JSGlobalObject* globalObject, Arra
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    if (JSArray* array = tryCreateContiguousArrayWithPattern(globalObject, pattern, count))
```
if (JSArray* array = ...; LIKELY(array))

hyjorc1 committed:
[JSC] Add fast paths for global atom RegExp matching one char and multiple chars
https://bugs.webkit.org/show_bug.cgi?id=281265
rdar://137717054

Reviewed by Yusuke Suzuki.

Inspired by [1], this patch add two fast paths to let global
atom RegExp to match strings quickly with simple string search.

[1] https://chromium-review.googlesource.com/c/v8/v8/+/5872644

                                         before                    after

global-atom-substring-match          2.2508+-0.0145     ^      1.6873+-0.0077        ^ definitely 1.3340x faster
global-atom-match-multi-chars        2.4767+-0.0195     ^      1.5488+-0.0071        ^ definitely 1.5991x faster
global-atom-match-one-char           8.0799+-0.0309     ^      1.2977+-0.0078        ^ definitely 6.2264x faster

* JSTests/microbenchmarks/global-atom-match-multi-chars.js: Copied from JSTests/microbenchmarks/substring-match-global-atom.js.
(test):
* JSTests/microbenchmarks/global-atom-match-one-char.js: Copied from JSTests/microbenchmarks/substring-match-global-atom.js.
(test):
* JSTests/microbenchmarks/global-atom-substring-match.js: Renamed from JSTests/microbenchmarks/substring-match-global-atom.js.
(test):
* JSTests/stress/string-16bit-repeat-overflow.js:
* JSTests/stress/substring-global-atom-cache.js:
(test):
* Source/JavaScriptCore/runtime/RegExpObject.cpp:
(JSC::RegExpObject::matchGlobal):
* Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
(JSC::GlobalAtomMeta::isSubstring):
(JSC::oneCharMatches):
(JSC::genericMatches):
(JSC::collectGlobalAtomMatches):
* Source/JavaScriptCore/runtime/RegExpSubstringGlobalAtomCache.cpp:
(JSC::RegExpSubstringGlobalAtomCache::tryGet):
(JSC::RegExpSubstringGlobalAtomCache::tryRecord):
(JSC::RegExpSubstringGlobalAtomCache::hasValidPattern): Deleted.
(JSC::RegExpSubstringGlobalAtomCache::collectMatches): Deleted.
* Source/JavaScriptCore/runtime/RegExpSubstringGlobalAtomCache.h:

Canonical link: https://commits.webkit.org/285090@main
